WordPress/gutenberg
# Fix Reset All button in the typography panel of the block inspector

## What?

The Reset All button in the "Typography" panel within block inspectors is not working as expected. This is a regression that we missed in the refactor from #47356 

## How?

The way "reset all" works is very weird: Each "ToolsPanelItem" registers a "filter" to apply when the "resetAll" is called and the wrapping `ToolsPanel` receives the registered filters as an argument to pass to its `resetAll` callback prop. 

The reason we have this instead of a regular callback, is because we want third-parties to be able to hook into the "resetAll" function call to register their own resets that can run in addition to the "core one". 

The way we register these callbacks now is by passing "registerAllFilter" prop to the `ToolsPanelItem` but the problem is that these filters assume that they receive a "block attributes" object and filter it. It makes it very hard to build a "reusable component" that is used in several contexts (inspector control of a given block and global styles) like we did with `TypographyPanel` and where we can't assume that we have an attributes object, all we have is a "style" object and "onChange" callback.

The solution here is a bit elaborate but allows us to have both:

 - I separated the registration of "resetAll" filters from the registration of "ToolsPanelItem" (the different controls within a panel) that way we can register a global "base" resetAll for all the TypographyPanel.
 - I've used the fact that we have the "as" prop to separate the generic case (global styles) from the specific case (block inspector) where we need to provide an alternative wrapper to inject in the block inspector. Within that specific "glue" component, I map the generic "resetAll" filter that filters a style object in to a "resetAl" filter that filters a block attributes object.

I couldn't think of any other solution, please let me know if you have any other ideas.

## Testing Instructions

1- Open the block inspector of a paragraph block
2- Edit some typography settings
3- Click "reset all" within the "..." menu.
4- It should reset all typography changes.
